packages/flutter/lib/src/rendering/object.dart

```diff
@@ -3248,30 +3249,87 @@ abstract class RenderObject with DiagnosticableTreeMixin implements HitTestTarge
-    final bool ancestorSpecified = ancestor != null;
     assert(attached);
-    if (ancestor == null) {
-      final RenderObject? rootNode = owner!.rootNode;
-      if (rootNode is RenderObject) {
-        ancestor = rootNode;
-      }
+    final Matrix4 identity = Matrix4.identity();
+    if (ancestor != null) {
+      return _computePaintTransform(ancestor, identity);
+    } else  {
+      final RenderObject? parent = this.parent;
+      return parent != null ? _computePaintTransformToGlobalCoordinates(parent, identity) : identity;
     }
-    final List<RenderObject> renderers = <RenderObject>[];
-    for (RenderObject renderer = this; renderer != ancestor; renderer = renderer.parent!) {
-      renderers.add(renderer);
-      assert(renderer.parent != null); // Failed to find ancestor in parent chain.
+  }
+
+  // Computes the paint transform from this to toRenderObject and applies it on
+  // `transform`.
+  Matrix4 _computePaintTransform(RenderObject toRenderObject, Matrix4 transform) {
+    if (!identical(this, toRenderObject)) {
+      final RenderObject? parent = this.parent;
+      assert(
+        toRenderObject.depth <= depth && parent != null,
+        '$toRenderObject is not an ancestor of the given RenderObject. Consider using computeTransformToRenderObject instead.',
+      );
+      parent!.applyPaintTransform(this, parent._computePaintTransform(toRenderObject, transform));
     }
-    if (ancestorSpecified) {
-      renderers.add(ancestor!);
+    return transform;
+  }
+
+  Matrix4 _computePaintTransformToGlobalCoordinates(RenderObject parent, Matrix4 transform) {
+    final RenderObject? parentParent = parent.parent;
+    if (parentParent != null) {
+      parent.applyPaintTransform(this, parent._computePaintTransformToGlobalCoordinates(parentParent, transform));
     }
+    assert(parentParent != null || parent is RenderView);
+    assert(parentParent == null || parent is! RenderView);
+    return transform;
+  }
+
+  /// Returns the paint transform that allows this `RenderObject` to perform
+  /// paint in `toRenderObject`'s coordinate system.
+  ///
+  /// Returns null if `toRenderObject` is not in the same render tree, or either
+  /// of the two `RenderObject`s is in an offscreen subtree.
+  ///
+  /// See also:
+  ///
+  ///  * [getTransformTo] which does the same, but only searches up the tree.
+  ///  * [paintsChild] which this method uses to determine whether
+  ///    a `RenderObject` is in an offscreen subtree.
+  Matrix4? computeTransformToRenderObject(RenderObject toRenderObject) {
     final Matrix4 transform = Matrix4.identity();
-    for (int index = renderers.length - 1; index > 0; index -= 1) {
-      renderers[index].applyPaintTransform(renderers[index - 1], transform);
+    final Matrix4 inverseTransform = Matrix4.identity();
+    if (_computePaintTranfrom(this, transform, toRenderObject, inverseTransform) && inverseTransform.invert() != 0) {
+      return inverseTransform..multiply(transform);
```

Comment:
Mathematically, this reads confusing. Doesn't multiplying the inverse matrix with the original matrix always give you the identity matrix by definition? So why not shortcut all this and just return `Matrix4.identity()`? 😄